runtime/plugin/osc52.lua

```diff
@@ -1,37 +1,72 @@
-for _, ui in ipairs(vim.api.nvim_list_uis()) do
-  if ui.chan == 1 and ui.stdout_tty then
-    tty = true
-    break
-  end
-end
-
--- Do not query when any of the following is true:
---   * TUI is not attached
---   * OSC 52 support is explicitly disabled via g:termfeatures
---   * Using a badly behaved terminal
-if
-  not tty
-  or (vim.g.termfeatures ~= nil and vim.g.termfeatures.osc52 == false)
-  or vim.env.TERM_PROGRAM == 'Apple_Terminal'
-then
-  return
-end
-
-require('vim.termcap').query('Ms', function(cap, found, seq)
-  if not found then
-    return
-  end
-
-  assert(cap == 'Ms')
-
-  -- If the terminal reports a sequence other than OSC 52 for the Ms capability
-  -- then ignore it. We only support OSC 52 (for now)
-  if not seq or not seq:match('^\027%]52') then
-    return
-  end
-
-  local termfeatures = vim.g.termfeatures or {}
-  termfeatures.osc52 = true
-  vim.g.termfeatures = termfeatures
-end)
+--- @class (private) TermFeatures
+--- @field osc52 boolean?
+
+local id = vim.api.nvim_create_augroup('nvim.osc52', { clear = true })
+vim.api.nvim_create_autocmd('UIEnter', {
+  group = id,
+  desc = 'Enable OSC 52 feature flag if a supporting TUI is attached',
+  callback = function()
+    -- If OSC 52 is explicitly disabled by the user then don't do anything
+    if vim.g.termfeatures ~= nil and vim.g.termfeatures.osc52 == false then
+      return
+    end
+
+    local tty = false
+    for _, ui in ipairs(vim.api.nvim_list_uis()) do
+      if ui.chan == 1 and ui.stdout_tty then
+        tty = true
+        break
+      end
+    end
+
+    -- Do not query when any of the following is true:
+    --   * No TUI is attached
+    --   * Using a badly behaved terminal
+    if not tty or vim.env.TERM_PROGRAM == 'Apple_Terminal' then
+      local termfeatures = vim.g.termfeatures or {} ---@type TermFeatures
+      termfeatures.osc52 = nil
+      vim.g.termfeatures = termfeatures
+      return
+    end
+
+    require('vim.termcap').query('Ms', function(cap, found, seq)
+      if not found then
+        return
+      end
+
+      assert(cap == 'Ms')
+
+      -- If the terminal reports a sequence other than OSC 52 for the Ms capability
+      -- then ignore it. We only support OSC 52 (for now)
+      if not seq or not seq:match('^\027%]52') then
+        return
+      end
+
+      local termfeatures = vim.g.termfeatures or {} ---@type TermFeatures
+      termfeatures.osc52 = true
+      vim.g.termfeatures = termfeatures
+    end)
+  end,
+})
+
+vim.api.nvim_create_autocmd('UILeave', {
+  group = id,
+  desc = 'Reset OSC 52 feature flag if no TUIs are attached',
+  callback = function()
+    -- If OSC 52 is explicitly disabled by the user then don't do anything
+    if vim.g.termfeatures ~= nil and vim.g.termfeatures.osc52 == false then
+      return
+    end
+
+    -- If no TUIs are attached then reset the OSC 52 term features flag
+    for _, ui in ipairs(vim.api.nvim_list_uis()) do
+      if ui.chan == 1 and ui.stdout_tty then
```

Comment:
the PR description mentions "check to make sure there is at is a TUI client connected to Nvim's stdout", but actually here we are checking specifically for chan=1.

Should the description say something like "check that a TUI is attached on channel 1"?

Reply:
The `chan == 1` check is probably not necessary actually, I think only `stdout_tty` is relevant here?